### PR TITLE
Add token_uri, on-chain metadata example

### PR DESCRIPTION
**Pull request by iLoveBusinessDevelopment**
Implements the changes described here: https://github.com/CosmWasm/cw-nfts/issues/2

`cw721_base` -- removes `name`, `description`, and `image`, only top level field is `token_uri` and `extension`. 
`cw721_metadata_onchain` -- example of Extension that moves metadata on-chain, supports name, description, image, and more

### CODE REVIEW

**Comment by the-frey** [comment]
At a skim, LGTM. Think the CircleCI config still needs updating however as `ci/circleci: contract_cw721_metadata_uri` has been removed AFAICT

**Comment by JakeHartnell** [comment]
Is there are reason the permissions on all the files changed?

**Comment by JakeHartnell** [comment]
Aside from the question about file permissions, this looks good to me.

**Comment by JakeHartnell** [comment]
Would be nice to validate the `token_uri` is indeed a valid URI and not a random string... but we can do this in a separate ticket if you wish.

**Comment by iLoveBusinessDevelopment** [comment]
> Is there are reason the permissions on all the files changed?

Oops, permission changes reverted.

> Would be nice to validate the `token_uri` is indeed a valid URI and not a random string... but we can do this in a separate ticket if you wish.

`token_uri` can be interpreted arbitrarily. Maybe you would try to check for http:// and https://, but now many NFTs use ipfs://. Something new can come out. Best just leave validating this up to the off-chain querier, who can change their validation logic to fit with the times.

**Review by JakeHartnell — APPROVED** [review]
LGTM. Thanks @alwin-peng!

**Comment by the-frey** [comment]
Agree with that. URIs are difficult to have a strong opinion on WRT validity and it's probably sensible to push that onto clients

**Comment by iLoveBusinessDevelopment** [comment]
gonna close and reopen to see if it fixes CI

**Comment by ethanfrey** [comment]
Just saw the CI issue... I had configured status checks in github. Fixed now.

I didn't get a chance to review today, but at least I could make your CI green

**Comment by the-frey** [comment]
What's left now before we merge this? Just Ethan's sense check?

File: contracts/cw721-metadata-onchain/src/lib.rs
Comment (ethanfrey) [inline contracts/cw721-metadata-onchain/src/lib.rs:28]: The only reason to use this over cw721-base is to use Metadata. I would make it required. (not an Option)

File: contracts/cw721-metadata-onchain/src/lib.rs
Comment (ethanfrey) [inline contracts/cw721-metadata-onchain/src/lib.rs:15]: As an example, this makes sense, that we store it in the exact same format that OpenSea has it.

However, if we build more complex NFTs, we would want to keep this `Metadata` struct as the query result from NFT Info (or have a custom query that just returns it). And maybe store it in a different format, that is more tied to the business case and just transform it to this format for display.

I am thinking Royalties. It is possible to store them in a Trait, but sub-optimal.
However, they could easily be converted to a Trait when serving a query

File: contracts/cw721-metadata-onchain/src/lib.rs
Comment (ethanfrey) [inline contracts/cw721-metadata-onchain/src/lib.rs:99]: This is a lot of `None`s... to save your fingers, I would:

1. `#[derive(Default)]` for Metadata and Trait
2. Auto-fill default fields:

```rust
Some(Metadata{
    description: Some("Spaceship with Warp Drive".into()),
    name: Some("Starship USS Enterprise".to_string()),
    ..Metadata::default()
})
```

Comment (ethanfrey) [reply contracts/cw721-metadata-onchain/src/lib.rs:99]: This doesn't change the function at all, but is nice to read and write

**Review by ethanfrey — APPROVED** [review]
Fine with me.

Added a few comments on code style you may want to integrate. But nothing blocking a merge

File: contracts/cw721-metadata-onchain/src/lib.rs
Comment (the-frey) [reply contracts/cw721-metadata-onchain/src/lib.rs:15]: I suppose there's also the question, since these are all Options anyway, whether we include the royalty fields for 2981 by default, and then simply leave it up to the implementer to define the custom queries that make up the rest of that royalty interface... 🤔

File: contracts/cw721-metadata-onchain/src/lib.rs
Comment (iLoveBusinessDevelopment) [reply contracts/cw721-metadata-onchain/src/lib.rs:28]: main reason is to interoperate with cw721-base better, so people can use `NftInfoResponse<Extension>` from `cw721-metadata-onchain` to query both cw721-base and cw721-metadata-onchain

File: contracts/cw721-metadata-onchain/src/lib.rs
Comment (iLoveBusinessDevelopment) [reply contracts/cw721-metadata-onchain/src/lib.rs:99]: will do in a later PR..

**Pull request merged**
